fuse_publishers/src/graph_dot_writer.cpp

```diff
@@ -0,0 +1,203 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2019, Clearpath Robotics
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <fuse_core/uuid.h>
+#include <fuse_msgs/Graph.h>
+#include <ros/ros.h>
+
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/graph/graphviz.hpp>
+
+#include <string>
+#include <vector>
+
+namespace fuse_publishers
+{
+
+class VertexProperty
+{
+public:
+  enum Type
+  {
+    VARIABLE,
+    CONSTRAINT,
+    UNKNOWN
+  };
+
+  VertexProperty(const std::string& uuid = fuse_core::uuid::to_string(fuse_core::uuid::NIL), const Type type = UNKNOWN,
+                 const std::string& subtype = "NO_SUBTYPE")
+    : uuid(uuid), type(type), subtype(subtype)
+  {
+  }
+
+  VertexProperty(const fuse_msgs::Variable& variable) : uuid(variable.uuid), type(VARIABLE), subtype(variable.type)
+  {
+  }
+
+  VertexProperty(const fuse_msgs::Constraint& constraint)
+    : uuid(constraint.uuid), type(CONSTRAINT), subtype(constraint.type)
+  {
+  }
+
+  std::string uuid;
+  Type type;
+  std::string subtype;
+};
```

Comment:
For now this is generic for `Variable` and `Constraint` classes, but we could split that.

For that reason I don't have the `data` from the variables here. It could be serialized into a string and added to the .dot file as well. But that's probably something we want to visualize with another tool. Here I'm thinking more about the graph topology, not the geometry of the the optimized variables.